File: examples/c-sharp-examples/MeshBoolean.dox.cs
public class MeshBooleanExample
{
    public static void Main(string[] args)
    {
        try
        {
            MR.Mesh mesh_a, mesh_b;
            if (args.Length >= 2)
            {
                // load the two meshes given on the command line
                mesh_a = MR.MeshLoad.fromAnySupportedFormat(args[0]);
                mesh_b = MR.MeshLoad.fromAnySupportedFormat(args[1]);
            }
            else
            {
                // no input given: make two unit spheres shifted along X, so that they overlap
                mesh_a = MR.makeUVSphere(1.0f, 64, 64);
                mesh_b = MR.makeUVSphere(1.0f, 64, 64);
                mesh_b.transform(MR.AffineXf3f.translation(new MR.Vector3f(0.7f, 0.0f, 0.0f)));
            }

            // optional mapper relating the primitives of the input meshes to the primitives of the result
            var mapper = new MR.BooleanResultMapper();
            var parameters = new MR.BooleanParameters();
            parameters.mapper = mapper;

            // perform boolean operation
            MR.BooleanResult res = MR.boolean(mesh_a, mesh_b, MR.BooleanOperation.Intersection, parameters);
            if (!res.valid())
            {
                Console.WriteLine("Error: {0}", res.errorString);
                return;
            }

            // find the faces of the result produced by each input mesh, and the faces the cut created
            var facesOfA = mapper.map(mesh_a.topology.getValidFaces(), MR.BooleanResultMapper.MapObject.A);
            var facesOfB = mapper.map(mesh_b.topology.getValidFaces(), MR.BooleanResultMapper.MapObject.B);
            var newFaces = mapper.newFaces();
            Console.WriteLine("faces from mesh A: {0}", facesOfA.count());
            Console.WriteLine("faces from mesh B: {0}", facesOfB.count());
            Console.WriteLine("faces created by the cut: {0}", newFaces.count());

            // map one particular face of mesh A forward: the cut can split it in several faces of
            // the result, or drop it completely if that part of mesh A is not in the result
            var faceOfA = new MR.FaceId(793);
            var oneFace = new MR.FaceBitSet(794);
            oneFace.set(faceOfA);
            var producedFaces = mapper.map(oneFace, MR.BooleanResultMapper.MapObject.A);
            Console.WriteLine("face {0} of mesh A produced {1} faces of the result", faceOfA.id, producedFaces.count());

            // and backward: the face of mesh A each face of the result came from
            // (invalid id for the faces that came from mesh B)
            if (producedFaces.count() > 0)
            {
                var new2OldFaces = mapper.getNew2OldFaceMap(MR.BooleanResultMapper.MapObject.A);
                var resultFace = producedFaces.find_first();
                Console.WriteLine("face {0} of the result came from face {1} of mesh A",
                    resultFace.id, new2OldFaces[resultFace].id);
            }

            // save result to STL file
            MR.MeshSave.toAnySupportedFormat(res.mesh, "out_boolean.stl");
        }
        catch (Exception e)
        {
            Console.WriteLine("Error: {0}", e.Message);
        }
    }
}

File: examples/js-examples/MeshBoolean.dox.js
import createMeshLib from '@meshinspector/meshlib';

const ml = await createMeshLib();

// create first sphere with radius of 1 unit
using sphere1 = ml.makeUVSphere(1.0, 64, 64);

// create second sphere (the JS bindings expose no cheap mesh copy, so regenerate an
// identical sphere) and move it in the X direction
using sphere2 = ml.makeUVSphere(1.0, 64, 64);
using xf = ml.AffineXf3f.translation({ x: 0.7, y: 0.0, z: 0.0 });
sphere2.transform(xf);

// optional mapper relating the primitives of the input meshes to the primitives of the result
using mapper = new ml.BooleanResultMapper();

// perform boolean operation
using result = ml.boolean(sphere1, sphere2, ml.BooleanOperation.Intersection, mapper);
if (!result.valid())
  throw new Error(result.errorString);

// find the faces of the result produced by each input sphere, and the faces the cut created
using topology1 = sphere1.topology;
using topology2 = sphere2.topology;
using validFaces1 = topology1.getValidFaces();
using validFaces2 = topology2.getValidFaces();
using facesOfSphere1 = mapper.mapFaces(validFaces1, ml.BooleanMapObject.A);
using facesOfSphere2 = mapper.mapFaces(validFaces2, ml.BooleanMapObject.B);
using newFaces = mapper.newFaces();
console.log(`faces from sphere1: ${facesOfSphere1.count()}`);
console.log(`faces from sphere2: ${facesOfSphere2.count()}`);
console.log(`faces created by the cut: ${newFaces.count()}`);

// map one particular face of sphere1 forward: the cut can split it in several faces of the
// result, or drop it completely if that part of sphere1 is not in the result
const faceOfSphere1 = 793;
using oneFace = ml.FaceBitSet.fromIndices([faceOfSphere1]);
using producedFaces = mapper.mapFaces(oneFace, ml.BooleanMapObject.A);
console.log(`face ${faceOfSphere1} of sphere1 produced ${producedFaces.count()} faces of the result`);

// and backward: the face of sphere1 each face of the result came from
// (4294967295, i.e. an invalid id, for the faces that came from sphere2)
using new2OldFaces = mapper.getNew2OldFaceMap(ml.BooleanMapObject.A);
const new2OldFacesArray = new2OldFaces.toArray();
const resultFace = producedFaces.find_first();
console.log(`face ${resultFace} of the result came from face ${new2OldFacesArray[resultFace]} of sphere1`);

// save result to STL file
using resultMesh = result.mesh;
ml.MeshSave.toAnySupportedFormat(resultMesh, 'out_boolean.stl');
